### ISSUE TITLE
Migrate warnings settings

### ISSUE THREAD

**Issue by snowleopard** [snowleopard/hadrian#296]
Warnings haven't been translated properly from the old build system to Hadrian, so we can sometimes generate lots and lots of them. I believe this is the reason behind [timing out on our OSX Travis instance](https://travis-ci.org/snowleopard/hadrian/jobs/189395066).

- [x] `-Wall` currently comes after package-specific warning settings coming from `cabal` files in the command line, overriding them.
- [x] General warnings settings: https://github.com/ghc/ghc/blob/master/mk/warnings.mk
- [x] RTS-related warnings: https://github.com/ghc/ghc/blob/master/rts/ghc.mk#L262
- [x] `ghc-cabal`-related warnings: https://github.com/ghc/ghc/blob/master/utils/ghc-cabal/ghc.mk#L55

**Comment by izgzhen** [comment]
What is the difference between "boot" and "install" in `XXX_dist-boot_EXTRA_HC_OPTS` and `XXX_dist-install_EXTRA_HC_OPTS` (https://github.com/ghc/ghc/blob/master/mk/warnings.mk)?

**Comment by izgzhen** [comment]
Oh never mind, I found that probably "boot" means stage1 and "install" means stage2

**Comment by izgzhen** [comment]
But why we don't add `Werror` anywhere in Hadrian now?

**Comment by izgzhen** [comment]
Some attempt:

```diff
diff --git a/src/Settings/Default.hs b/src/Settings/Default.hs
index d242502..f5bf97f 100644
--- a/src/Settings/Default.hs
+++ b/src/Settings/Default.hs
@@ -215,6 +215,12 @@ defaultBuilderArgs = mconcat
     , makeBuilderArgs
     , tarBuilderArgs ]
 
+-- | Disable some warnings in packages we use
+-- | https://github.com/ghc/ghc/blob/master/mk/warnings.mk#L46
+disableWarningArgs :: Args
+disableWarningArgs = stage Stage0 ? builder Ghc ? mconcat
+    [ package transformers ? mconcat [ arg "-fno-warn-unused-matches", arg "-fno-warn-unused-imports" ] ]
+
 -- | All 'Package'-dependent command line arguments.
 defaultPackageArgs :: Args
 defaultPackageArgs = mconcat
@@ -227,4 +233,5 @@ defaultPackageArgs = mconcat
     , haddockPackageArgs
     , integerGmpPackageArgs
     , rtsPackageArgs
-    , runGhcPackageArgs ]
+    , runGhcPackageArgs
+    , disableWarningArgs ]
```

**Comment by snowleopard** [comment]
> Oh never mind, I found that probably "boot" means stage1 and "install" means stage2

@izgzhen More accurately, `boot` means `Stage0` in Hadrian, and `install` means either `Stage1` or `Stage2` (this was not consistent in Make).

> But why we don't add Werror anywhere in Hadrian now?

I guess only because we'll get a lot of errors. But we should have it!

> Some attempt:

Is this attempt for `transformers` only? Presumably there are many more warning flags to add.

But this is the right step forward 👍

**Comment by izgzhen** [comment]
> Is this attempt for transformers only? Presumably there are many more warning flags to add. But this is the right step forward

Awesome. I will finish the rest of it later :)

**Comment by izgzhen** [comment]
Report something I found after switching on `-Werror` experimentally (on macOS):

```
/usr/bin/gcc -m64 -fno-stack-protector -Iincludes -I_build/generated -I_build/stage1/libraries/ghc-prim -I/Users/zz/Repos/ghc/_build/stage1/rts -I/Users/zz/Repos/ghc/includes -I/Users/zz/Repos/ghc/_build/generated -Werror -c libraries/ghc-prim/cbits/atomic.c -o _build/stage1/libraries/ghc-prim/c/cbits/atomic.o -Wno-sync-nand

error: unknown warning option '-Wno-sync-nand' [-Werror,-Wunknown-warning-option]
```

This is caused by interpretation of `GccLt44` -- on macOS (Argh!), `gcc` is actually `clang`, which doesn't know `-Wno-sync-nand`. And this will halt the build when `-Werror` is enabled.

**Comment by izgzhen** [comment]
I can fix the above problem with `(not <$> flag GccIsClang)`

**Comment by snowleopard** [comment]
Yes, `not <$> flag GccIsClang` is the right way. As far as I can see this corresponds to this code in Make:
```
ifneq "$(GccIsClang)" "YES"

# Debian doesn't turn -Werror=unused-but-set-variable on by default, so
# we turn it on explicitly for consistency with other users
ifeq "$(GccLT46)" "NO"
# Never set the flag on Windows as the host gcc may be too old.
ifneq "$(HostOS_CPP)" "mingw32"
SRC_CC_WARNING_OPTS += -Werror=unused-but-set-variable
endif
endif

ifeq "$(GccLT44)" "NO"
# Suppress the warning about __sync_fetch_and_nand (#9678).
libraries/ghc-prim/cbits/atomic_CC_OPTS += -Wno-sync-nand
# gcc 4.6 gives 3 warnings for giveCapabilityToTask not being inlined
# gcc 4.4 gives 2 warnings for lockClosure not being inlined
SRC_CC_WARNING_OPTS += -Wno-error=inline
endif

else

# Don't warn about unknown GCC pragmas when using clang
SRC_CC_WARNING_OPTS += -Wno-unknown-pragmas

endif
```

**Comment by izgzhen** [comment]
Yes you are right -- I am now actually confused by `SRC_CC_WARNING_OPTS` now -- it is used in `includes/ghc.mk`, which seems to be something related to linking. Are you familiar with this make variable?

**Comment by snowleopard** [comment]
According to this comment in `mk/config.mk` this variable is exactly for suppressing warnings:
```
# Warning suppression flags. See mk/warnings.mk.
SRC_CC_WARNING_OPTS =
SRC_HC_WARNING_OPTS =
```
If I remember correctly, `includes/ghc.mk` is responsible for generating included header files.

**Comment by izgzhen** [comment]
Yes, true. But which builder will this option affect? What is the difference between `SRC_CC_WARNING_OPTS` and `SRC_CC_OPTS`?

**Comment by snowleopard** [comment]
I believe `SRC_CC_OPTS` are base flags for C compiler. Various other variables, including `SRC_CC_WARNING_OPTS` are used to override these base flags.

However, I don't know whether there is comprehensive documentation about this -- I mostly had to re-engineer what's going on by looking at comments and talking to various people. (So, I may be wrong too!)

**Comment by snowleopard** [comment]
@izgzhen What is the status of this? My feeling was that you've already fixed all missing warnings, so is the above todo list out-of-date?

**Comment by izgzhen** [comment]
- https://github.com/snowleopard/hadrian/blob/master/src/Settings/Builders/Cc.hs#L16-L17
- https://github.com/ghc/ghc/blob/master/mk/warnings.mk#L16-L18
- https://github.com/ghc/ghc/blob/master/mk/warnings.mk#L46-L47
- https://github.com/ghc/ghc/blob/master/mk/warnings.mk#L73
- https://github.com/ghc/ghc/blob/master/mk/warnings.mk#L78

**Comment by izgzhen** [comment]
please wait a bit, I am still updating the list, but

1. I believed that there must be something missing
2. But maybe some of them in `Makefile`s are not really necessary? considering the difficulty of maintaining the validity of these flags

**Comment by snowleopard** [comment]
@izgzhen Thanks! Can you tick off the items in the top todo list? If not, let me know what to mark as complete and I'll do that. 

> But maybe some of them in `Makefiles` are not really necessary? considering the difficulty of maintaining the validity of these flags

It's quite possible, of course. If you don't know what a particular warning flag is for, I suggest we skip it. If it is actually needed someone knowledgeable will add it to Hadrian later.

**Comment by izgzhen** [comment]
> Thanks! Can you tick off the items in the top todo list? If not, let me know what to mark as complete and I'll do that.

It looks like all three unmarked categories has one or more lines as in the latest list just collected by me :( so I am not sure if any of them can be really ticked off

**Comment by snowleopard** [comment]
@izgzhen Let me take this over this. I have some time today, so hopefully will be able to fix the remaining warnings. Here is the todo list based on your findings:

- [ ] https://github.com/snowleopard/hadrian/blob/master/src/Settings/Builders/Cc.hs#L16-L17
- [ ] https://github.com/ghc/ghc/blob/master/mk/warnings.mk#L16-L18
- [ ] https://github.com/ghc/ghc/blob/master/mk/warnings.mk#L46-L47
- [ ] https://github.com/ghc/ghc/blob/master/mk/warnings.mk#L73
- [ ] https://github.com/ghc/ghc/blob/master/mk/warnings.mk#L78
- [ ] https://github.com/ghc/ghc/blob/master/mk/warnings.mk#L104
- [ ] https://github.com/ghc/ghc/blob/master/mk/warnings.mk#L110-L114
- [ ] https://github.com/ghc/ghc/blob/master/rts/ghc.mk#L303-L321
- [ ] https://github.com/ghc/ghc/blob/master/utils/ghc-cabal/ghc.mk#L63

**Comment by snowleopard** [comment]
I think I've fixed the warnings in #448, so we can close this.

A couple of notes:
* There were a few Windows-specific warning suppressions, but they appear to be outdated (I don't get any warnings on my Windows machine).
* I didn't include warning suppression flags for packages that we do not build (such as `vector`).
* Apparently, `Cc` builder is unused at the moment -- all C files are compiled with GHC. 
* I dropped `-Wno-tabs` on `xhtml` package -- there appear to be no tabs in it.
